### PR TITLE
ffh.dhcp_server: switch to Kea DHCP server

### PR DESCRIPTION
**Pull request by CodeFetch**
Kea is the successor of ISC DHCP and will hopefully not cause
problems as like issue #109.

### CODE REVIEW

File: roles/ffh.dhcp_server/tasks/main.yml [hunk 1]
```diff
@@ -1,25 +1,23 @@
...
 
-- name: Install dhcpd-pools
-  apt: name=dhcpd-pools
+- name: Install kea-dhcp4-server
+  apt: name=kea-dhcp4-server update_cache=yes
   
-- name: Ramdisk in fstab eintragen
-  notify: Restart isc-dhcp-server
-  mount: name=/var/lib/dhcp src=tmpfs fstype=tmpfs opts='defaults,size=100M' state=mounted
+- name: Add a ramdisk entry in fstab
```
Comment (1977er) [inline roles/ffh.dhcp_server/tasks/main.yml:6]: Although frequently rebooting a supernode is not recommened, we are losing dhcp state, if we put everything on a ramdisk. If we use ramdisks we have to make sure, that lease candidate addresses are probed properly.

Not using a ramdisk has privacy implications.

File: roles/ffh.dhcp_server/templates/dhcpd.conf.j2 [hunk 2]
```diff
@@ -1,17 +1,70 @@
...
+	// when they get their own client-id. Kea can disable RFC6842 support.
+	"echo-client-id": false,
+
+	// use MAC address for IP address calculation
+	"match-client-id": false,
+
+	"loggers": [{
+		"name": "kea-dhcp4",
+		"output_options": [{ "output": "syslog:dhcpd", }],
+		"severity": "WARN"
```
Comment (1977er) [inline roles/ffh.dhcp_server/templates/dhcpd.conf.j2:46]: What does WARN include? Do we store more information than we have to with this feature?

File: roles/ffh.dhcp_server/tasks/per_domain.yml [hunk 3]
```diff
@@ -1,8 +1,10 @@
...
+---
+
+- name: Set domain-specific variables
   set_fact:
     dhcp_net:
       net: "10.{{ domain.id }}.0.0"
-      mask: "255.255.0.0"
+      cidr: 16
     dhcp_interface: "bat{{ domain.id }}"
     dhcp_range:
```
Comment (1977er) [inline roles/ffh.dhcp_server/tasks/per_domain.yml:9]: Maybe we should rethink our decision to use decimal-oriented boundaries for our ranges. I would prefer ranges which can simply be represented in a cidr notation.

**Comment by CodeFetch** [comment]
Closed as Kea is badly maintained in Debian:
https://tracker.debian.org/pkg/isc-kea

**Comment by 1977er** [comment]
Why do we care? We are using other software that is not even available from Debian. Or maintained at all.

**Comment by AiyionPrime** [comment]
I don't think this should stop us.
In case we did care, we might even better help the maintainer out.

**Comment by AiyionPrime** [comment]
I've looked into it;
this does not even look like the maintainers were sleeping at all, but as if debians release cycle is just awfully slow.
I'd like to reopen this. Version 1.1 would be fine for me for now, its security record is not that bad.
Version 1.4.0 has had a leak but that's taken care of already.

https://qa.debian.org/excuses.php?package=isc-kea

**Comment by AiyionPrime** [comment]
In case we wanted to kick packages this one would be a good start.
https://tracker.debian.org/pkg/log4cplus

**Comment by CodeFetch** [comment]
I'm against choosing Kea and will not invest more time into it. That's why I closed it. If you want to support Kea, feel free to take over the PR. I've reformulated @Manawyrm pros/cons to reflect my experience.

Kea DHCP: 
- badly-written documentation (steep learning curve/hard to find information)
+ JSON-style
- JSON-format is not well thought-through, because aim was to mimic ISC config format
+ modern
- not included in a recent version in Debian stable
# pretends to handle large setups fine
# pretends to be a professional enterprise-solution
- not so interesting for open-source developers (therefore potentially badly reviewed)
- the debian-shipped version is so old it doesn't even allow comments in config files

dnsmasq:
+ has the most active developer community
+ we know some of the contributors as it is the standard on OpenWrt and will therefore get better support when we have issues
# Manawyrm says does not handle large installation well, but I can't find any issues by other people
- bad security track record
+ someone cared enough to fix security

ddhcpd: 
+ runs directly on the gluon nodes
+ quick response times
- needs secondary fallback DHCP for old nodes/without autoupdater
- does likely not include optimizations for increasing compatibility with e.g. Windows
- needs testing

**Comment by CodeFetch** [comment]
It just feels wrong. Hope you understand, that I don't have any motivation to try Kea anymore. Maybe sometime in the future, when there is a stable release in Debian.

**Pull request closed without merging**